### ISSUE TITLE
--auto-gen-config sets a value to "-Infinity" for WordArray

### ISSUE THREAD

**Issue by maxjacobson** [bbatsov/rubocop#2740]
I haven't investigated exactly why this is occurring yet, but running `rubocop --auto-gen-config` with Rubocop 0.36 on a codebase generated this config:

``` yaml
Style/WordArray:
  EnforcedStyle: percent
  MinSize: -Infinity
```

And this config is invalid and leads to errors later when it's loaded. I don't think our word arrays are particularly unusual -- we have 4 of them, ranging in length from 2 to 3 elements.

**Comment by jonas054** [comment]
I think we need more help with reproducing. Do you use square brackets or `%w` in the word arrays you have? What's your configured style before running `rubocop --auto-gen-config`?

**Comment by maxjacobson** [comment]
Using square brackets before

The earlier style (from an earlier run of `rubocop --auto-gen-config`) was:

```yaml
# Offense count: 2                                                                  
# Cop supports --auto-correct.                                                      
# Configuration parameters: WordRegex.                                              
Style/WordArray:                                                                    
  MinSize: 3                                                                        
```

**Comment by jonas054** [comment]
Thanks. I still haven't been able to reproduce the problem, though. 

The values you see in the generated configuration must come from [these assignments](https://github.com/bbatsov/rubocop/blob/v0.36.0/lib/rubocop/cop/style/word_array.rb#L120-L121), but I can't figure out what sequence of calls to the `style_detected` method could lead to this outcome.

@maxjacobson Can you add `p style, ary_size` at the top of `style_detected` and then reproduce the problem? That might bring some clarity.

@alexdowad Do you have any better ideas?

**Comment by madwort** [comment]
Is there any chance that https://github.com/bbatsov/rubocop/blob/v0.36.0/lib/rubocop/cop/style/word_array.rb#L25 `node.children` might not be an array & be some other object that also implements `.size` but gives a weird result (if was an object whose `.size` returned `Float::NAN` that might do it)?

**Comment by alexdowad** [comment]
@jonas054 `@largest_brackets` is initialized to negative infinity on line 104. That is so that any value found in any source file will always be larger. The intention was for that value to always be overwritten on the very first call to this method, but apparently it's not being overwritten.

**Comment by alexdowad** [comment]
The only reason why the initial value would never be overwritten is if `array` style is never detected.

**Comment by alexdowad** [comment]
@maxjacobson Please edit `word_array.rb` and add the following line after line 102:

```ruby
puts "**** detected style: #{style} array size: #{ary_size} config_to_allow_offenses: #{cfg.inspect}"
```

Then try again and post the console output here.

**Comment by maxjacobson** [comment]
![screenshot from 2016-02-03 08 24 55](https://cloud.githubusercontent.com/assets/1421211/12783438/a25e1958-ca4f-11e5-9641-d698d33f69fc.png)

was just adding some logging :smile: 

it looks like it transitions from brackets to percent and that's when it breaks...

**Comment by alexdowad** [comment]
The issue is that a new cop object is instantiated for each inspected file, but the `config_to_allow_offenses` persists.

**Comment by alexdowad** [comment]
Let me see what I can do about this. @jonas054 you could assign this issue to me if you like.

**Comment by maxjacobson** [comment]
Sounds like you know what's up, but here's the output in the requested format anyway)

```
# Logfile created on 2016-02-03 08:30:27 -0500 by logger.rb/41954
I, [2016-02-03T08:30:27.811472 #30703]  INFO -- : **** detected style: brackets array size: 3 config_to_allow_offenses: {}
I, [2016-02-03T08:30:29.099091 #30703]  INFO -- : **** detected style: brackets array size: 2 config_to_allow_offenses: {"EnforcedStyle"=>"brackets"}
I, [2016-02-03T08:30:29.541292 #30703]  INFO -- : **** detected style: brackets array size: 3 config_to_allow_offenses: {"EnforcedStyle"=>"brackets"}
I, [2016-02-03T08:30:30.045335 #30703]  INFO -- : **** detected style: brackets array size: 3 config_to_allow_offenses: {"EnforcedStyle"=>"brackets"}
I, [2016-02-03T08:30:30.177833 #30703]  INFO -- : **** detected style: percent array size: 1 config_to_allow_offenses: {"EnforcedStyle"=>"brackets"}
I, [2016-02-03T08:30:30.179034 #30703]  INFO -- : **** detected style: percent array size: 4 config_to_allow_offenses: {"EnforcedStyle"=>"percent", "MinSize"=>-Infinity}
I, [2016-02-03T08:30:30.215578 #30703]  INFO -- : **** detected style: percent array size: 1 config_to_allow_offenses: {"EnforcedStyle"=>"percent", "MinSize"=>-Infinity}
I, [2016-02-03T08:30:30.216600 #30703]  INFO -- : **** detected style: percent array size: 1 config_to_allow_offenses: {"EnforcedStyle"=>"percent", "MinSize"=>-Infinity}
I, [2016-02-03T08:30:30.230177 #30703]  INFO -- : **** detected style: percent array size: 1 config_to_allow_offenses: {"EnforcedStyle"=>"percent", "MinSize"=>-Infinity}
I, [2016-02-03T08:30:30.231207 #30703]  INFO -- : **** detected style: percent array size: 1 config_to_allow_offenses: {"EnforcedStyle"=>"percent", "MinSize"=>-Infinity}
I, [2016-02-03T08:30:30.243237 #30703]  INFO -- : **** detected style: percent array size: 1 config_to_allow_offenses: {"EnforcedStyle"=>"percent", "MinSize"=>-Infinity}
I, [2016-02-03T08:30:30.244243 #30703]  INFO -- : **** detected style: percent array size: 1 config_to_allow_offenses: {"EnforcedStyle"=>"percent", "MinSize"=>-Infinity}
```

**Comment by jonas054** [comment]
@alexdowad Thanks for stepping up! I can't find a way to assign it to you in the GitHub UI (only to the four people who are collaborators), but let it be known that I hereby hand it over to you.

**Comment by jaredbeck** [comment]
If it helps, I can confirm that `--auto-gen-config` still sets `-Infinity` in 0.42.0.

**Comment by backus** [comment]
Here is a minimal reproduction:

```sh
$ cat a.rb
['g', 'h']

$ cat c.rb
%w(a b)

$ ls -la
total 16
drwxr-xr-x   4 johnbackus  staff   136B Jan 24 19:09 .
drwxr-xr-x@ 52 johnbackus  staff   1.7K Jan 24 19:03 ..
-rw-r--r--   1 johnbackus  staff    11B Jan 24 19:07 a.rb
-rw-r--r--   1 johnbackus  staff     8B Jan 24 19:07 c.rb

$ rubocop -V
0.47.1 (using Parser 2.3.3.1, running on ruby 2.3.1 x86_64-darwin16)

$ rubocop --auto-gen-config a.rb c.rb
Inspecting 2 files
C.

Offenses:

a.rb:1:1: C: Use %w or %W for an array of words.
['g', 'h']
^^^^^^^^^^

2 files inspected, 1 offense detected
Created .rubocop_todo.yml.
Run `rubocop --config .rubocop_todo.yml`, or add `inherit_from: .rubocop_todo.yml` in a .rubocop.yml file.

$ cat .rubocop_todo.yml
# This configuration was generated by
# `rubocop --auto-gen-config`
# on 2017-01-24 19:10:11 -0800 using RuboCop version 0.47.1.
# The point is for the user to remove these configuration records
# one by one as the offenses are removed from the code base.
# Note that changes in the inspected code, or installation of new
# versions of RuboCop, may require this file to be generated again.

# Offense count: 1
# Cop supports --auto-correct.
# Configuration parameters: SupportedStyles, WordRegex.
# SupportedStyles: percent, brackets
Style/WordArray:
  EnforcedStyle: percent
  MinSize: -Infinity

**Comment by mikegee** [comment]
I successfully duplicated this following @backus's procedure, but I'm having trouble translating that into a failing spec. [Here's my attempt so far.](https://github.com/mikegee/rubocop/commit/8bbfd5292c1641058e1f3fcccabf9508c584eb29) I hacked in some helpers that were supposed to simulate processing multiple files. I wonder if there are any existing specs that process multiple files I could use for guidance.

Those specs are giving `"Enabled" => false` rather than `"MinSize" => -Infinity`.

**Comment by jaredbeck** [comment]
I highly recommend reading the commit message on https://github.com/bbatsov/rubocop/commit/9a96a58cd63735cc001ca6a8fc612b94002f1ea4 lol
